logins-sql/examples/sync_pass_sql.rs

```diff
@@ -0,0 +1,502 @@
+                Value::Integer(i) => Cell::new(&i.to_string()).style_spec("Fb"),
+                Value::Real(r) => Cell::new(&r.to_string()).style_spec("Fb"),
+                Value::Text(s) => Cell::new(&s.to_string()).style_spec("Fr"),
+                Value::Blob(b) => Cell::new(&format!("{}b blob", b.len()))
+            }
+        }).collect::<Vec<_>>()
+    })?;
+
+    for row in rows {
+        table.add_row(Row::new(row?));
+    }
+    table.printstd();
+    Ok(())
+}
+
+fn show_all(engine: &PasswordEngine) -> Result<Vec<String>> {
+    let records = engine.list()?;
+    let last_sync = engine.db.get_last_sync()?.unwrap_or_default();
+    let changes = engine.db.fetch_outgoing(last_sync)?;
+    info!("Engine has {} records, a last sync timestamp of {}, and {} queued changes",
+           records.len(), last_sync, changes.changes.len());
+
+    let mut table = prettytable::Table::new();
+
+    table.add_row(row![bc =>
+        "(idx)",
+        "Guid",
+        "Username",
+        "Password",
+        "Host",
+
+        "Submit URL",
+        "HTTP Realm",
+
+        "User Field",
+        "Pass Field",
+
+        "Uses",
+        "Created At",
+        "Changed At",
+        "Last Used"
+    ]);
+
+    let mut v = Vec::with_capacity(records.len());
+    let mut record_copy = records.clone();
+    record_copy.sort_by(|a, b| a.id.cmp(&b.id));
+    for rec in records.iter() {
+        table.add_row(row![
+            r->v.len(),
+            Fr->&rec.id,
+            &rec.username,
+            Fd->&rec.password,
+
+            &rec.hostname,
+            string_opt_or(&rec.form_submit_url, ""),
+            string_opt_or(&rec.http_realm, ""),
+
+            &rec.username_field,
+            &rec.password_field,
+
+            rec.times_used,
+            timestamp_to_string(rec.time_created),
+            timestamp_to_string(rec.time_password_changed),
+            timestamp_to_string(rec.time_last_used)
+        ]);
+        v.push(rec.id.clone());
+    }
+    table.printstd();
+    Ok(v)
+}
+
+fn prompt_record_id(e: &PasswordEngine, action: &str) -> Result<Option<String>> {
+    let index_to_id = show_all(e)?;
+    let input = if let Some(input) = prompt_usize(&format!("Enter (idx) of record to {}", action)) {
+        input
+    } else {
+        return Ok(None);
+    };
+    if input >= index_to_id.len() {
+        info!("No such index");
+        return Ok(None);
+    }
+    Ok(Some(index_to_id[input].clone()))
+}
+
+fn init_logging() {
+    // Explicitly ignore some rather noisy crates. Turn on trace for everyone else.
+    let spec = "trace,tokio_threadpool=warn,tokio_reactor=warn,tokio_core=warn,tokio=warn,hyper=warn,want=warn,mio=warn,reqwest=warn";
+    env_logger::init_from_env(
+        env_logger::Env::default().filter_or("RUST_LOG", spec)
+    );
+}
+
+fn main() -> Result<()> {
+    init_logging();
+    std::env::set_var("RUST_BACKTRACE", "1");
+
+    let matches = clap::App::new("sync_pass_sql")
```

Comment:
It is. I also used it [in places-tool](https://github.com/thomcc/places-tool/blob/master/src/main.rs#L26-L65).

That said, after writing this I learned that docopt.rs is by BurntSushi, the author of the regex crate, ripgrep, among several other very high quality tools (this is probably a bad way to judge library quality, but whatever).

So I might try that one out next time? I had avoided it because the documentation was confusing to me since a lot of the work is parsed from `usage` comments.